### PR TITLE
[WIP] optional_unbox

### PR DESCRIPTION
**Pull request by bobzhang**

### CODE REVIEW

**Comment by cristianoc** [comment]
How about removing the "opam" part form the tests, it's pretty much only noise.
```
 478 test(s) passed
    0 test(s) failed
    0 unexpected error(s)
```

File: bytecomp/matching.ml [hunk 1]
```diff
@@ -1343,13 +1343,33 @@ let matcher_constr cstr = match cstr.cstr_arity with
...
+     prim_arity = 1;
+     prim_alloc = false;
+     prim_native_name = "";
+     prim_native_float = false}
+    
+let cstr_is_option
+    ({cstr_res = {desc ; _}} : Types.constructor_description) =
+  match desc with
+  | Tconstr(path_option,_,_) -> 
+      Path.same path_option Predef.path_option
```
Comment (bobzhang) [inline bytecomp/matching.ml]: maybe not enough due to type alias?

Comment (bobzhang) [reply bytecomp/matching.ml]: but it is worth invesigating, since it may give some good enough info to know it is not an option type?

Comment (cristianoc) [reply bytecomp/matching.ml]: I think this is tight enough.
I thought the type information would give us the list and arity of all the other constructors. Apparently it does not. We could add it, but it seems not worth it.

**Pull request closed without merging**